examples/acoustic/wavesolver.py

```diff
@@ -0,0 +1,224 @@
+        return ForwardOperator(self.model, save=True, source=self.source,
+                               receiver=self.receiver, time_order=self.time_order,
+                               space_order=self.space_order, **self._kwargs)
+
+    @property
+    def op_adj(self):
+        """Cached operator for adjoint runs"""
+        return AdjointOperator(self.model, save=False, source=self.source,
+                               receiver=self.receiver, time_order=self.time_order,
+                               space_order=self.space_order, **self._kwargs)
+
+    @property
+    def op_grad(self):
+        """Cached operator for gradient runs"""
+        return GradientOperator(self.model, save=False, source=self.source,
+                                receiver=self.receiver, time_order=self.time_order,
+                                space_order=self.space_order, **self._kwargs)
+
+    @property
+    def op_born(self):
+        """Cached operator for born runs"""
+        return BornOperator(self.model, save=False, source=self.source,
+                            receiver=self.receiver, time_order=self.time_order,
+                            space_order=self.space_order, **self._kwargs)
+
+    def forward(self, src=None, rec=None, u=None, m=None, save=False, **kwargs):
+        """
+        Forward modelling function that creates the necessary
+        data objects for running a forward modelling operator.
+
+        :param src: Symbol with time series data for the injected source term
+        :param rec: Symbol to store interpolated receiver data
+        :param u: (Optional) Symbol to store the computed wavefield
+        :param m: (Optional) Symbol for the time-constant square slowness
+        :param save: Option to store the entire (unrolled) wavefield
+
+        :returns: Receiver, wavefield and performance summary
+        """
+        # Source term is read-only, so re-use the default
+        if src is None:
+            src = self.source
+        # Create a new receiver object to store the result
+        if rec is None:
+            rec = Receiver(name='rec', ntime=self.receiver.nt,
+                           coordinates=self.receiver.coordinates.data)
+
+        # Create the forward wavefield if not provided
+        if u is None:
+            u = TimeData(name='u', shape=self.model.shape_domain,
+                         save=save, time_dim=self.source.nt,
+                         time_order=self.time_order,
+                         space_order=self.space_order,
+                         dtype=self.model.dtype)
+
+        # Pick m from model unless explicitly provided
+        if m is None:
+            m = m or self.model.m
+
+        # Execute operator and return wavefield and receiver data
+        if save:
+            summary = self.op_fwd_save.apply(src=src, rec=rec, u=u, m=m, **kwargs)
+        else:
+            summary = self.op_fwd.apply(src=src, rec=rec, u=u, m=m, **kwargs)
+        return rec, u, summary
```

Comment:
Let's see if I get this right: it'd make no sense to subclass``AcousticWaveSolver`` with a class that implicitly has ``save = True`` because in general we'll need to apply Forward in both ways, for the same solver, right?